File: LeetCode/go/LC0047_Permutations_II/permutations_ii.go
package permutations_ii

import "sort"

func permuteUnique(nums []int) (res [][]int) {
	var path []int
	n := len(nums)
	visited := make([]bool, n)

	var backtrack func(int)
	backtrack = func(cur int) {
		if cur == n {
			res = append(res, append([]int{}, path...))
			return
		}

		for i := 0; i < n; i++ {
			if visited[i] || (i > 0 && nums[i] == nums[i-1] && !visited[i-1]) {
				continue
			}
			visited[i] = true
			path = append(path, nums[i])
			backtrack(cur + 1)
			path = path[:len(path)-1]
			visited[i] = false
		}
	}

	sort.Ints(nums)
	backtrack(0)
	return res
}

File: LeetCode/java/LC0047_Permutations_II/PermutationsII.java
package LC0047_Permutations_II;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * <a href="https://leetcode.cn/problems/permutations-ii/">Permutations II</a>
 * 数组；回溯；排序
 */
public class PermutationsII {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.permuteUnique(new int[]{1, 1, 2}));
        System.out.println(solution.permuteUnique(new int[]{1, 2, 3}));
    }
}

class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        boolean[] visited = new boolean[nums.length];

        Arrays.sort(nums);
        backtrack(0, nums, path, visited, res);
        return res;
    }

    private void backtrack(int cur, int[] nums, List<Integer> path, boolean[] visited, List<List<Integer>> res) {
        if (cur == nums.length) {
            res.add(new ArrayList<>(path));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (visited[i] || (i > 0 && nums[i] == nums[i - 1] && !visited[i - 1])) {
                continue;
            }
            visited[i] = true;
            path.add(nums[i]);
            backtrack(cur + 1, nums, path, visited, res);
            path.remove(path.size() - 1);
            visited[i] = false;
        }
    }
}